lib/auth/oidc_google.go

```diff
@@ -0,0 +1,214 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/utils"
+
+	"github.com/coreos/go-oidc/jose"
+	"github.com/gravitational/trace"
+	"golang.org/x/oauth2/google"
+	directory "google.golang.org/api/admin/directory/v1"
+	"google.golang.org/api/cloudidentity/v1"
+	"google.golang.org/api/option"
+)
+
+const (
+	// gsuiteIssuerURL is the issuer URL for Google Workspace accounts.
+	gsuiteIssuerURL = "https://accounts.google.com"
+
+	// gsuiteGroupsClaim is the OIDC claim that we inject into the claims
+	// returned for Google Workspace users, containing the email addresses of
+	// the Google Groups that the user belongs to.
+	gsuiteGroupsClaim = "groups"
+)
+
+// addGsuiteClaims will fetch extra data from proprietary Google APIs if
+// applicable to the connector, and it will add claims based on the fetched
+// data. The current implementation adds a "groups" claim containing the Google
+// Groups that the user is a member of.
+//
+// If clientOptions is not empty, it will be passed to the underlying google api
+// client without loading the credentials according to the configuration of the
+// connector. The credentials in the default case are loaded from the connector
+// with the getGsuiteCredentialsForConnector function.
+func addGsuiteClaims(ctx context.Context, connector types.OIDCConnector, claims jose.Claims, clientOptions ...option.ClientOption) (jose.Claims, error) {
+	// If google_service_account_uri and google_service_account are not set, we
+	// assume that this is a non-GWorkspace OIDC provider using the same
+	// issuer URL as Google Workspace (e.g.
+	// https://developers.google.com/identity/protocols/oauth2/openid-connect).
+	if connector.GetIssuerURL() != gsuiteIssuerURL || (connector.GetGoogleServiceAccountURI() == "" && connector.GetGoogleServiceAccount() == "") {
+		return claims, nil
+	}
```

Comment:
I'd wrap this into it's own function `isGoogleWorkspace` then you can update `getClaims` to have something like the following:

```go
// If the OIDC provider is configured to be Google Workspace, add in addition claims.
if ok := isGoogleWorkspace(...); ok {
    claims, err = addGoogleWorkspaceClaims(...)
    if err != nil {
        return nil, trace.Wrap(err)
    }
}
return claims, nil
```